packages/itwinui-react/src/core/InputGrid/InputGrid.tsx

```diff
@@ -49,4 +70,207 @@ export const InputGrid = React.forwardRef((props, ref) => {
 
 //-------------------------------------------------------------------------------
 
+/**
+ * Ensures that label, input and message are properly associated
+ * with each other, for accessibility purposes.
+ *
+ * - `Select` will be associated with label using `aria-labelledby`
+ * - Other inputs will be associated with label using `htmlFor`
+ * - Message will be associated with input/select using `aria-describedby`
+ */
+const useChildrenWithIds = (children: React.ReactNode) => {
+  const { labelId, inputId, messageId } = useSetup(children);
+
+  return React.useMemo(
+    () =>
+      React.Children.map(children, (child) => {
+        if (!React.isValidElement(child)) {
+          return child;
+        }
+
+        if (child.type === Label || child.type === 'label') {
+          return cloneElementWithRef(child, (child) => ({
+            ...child.props,
+            htmlFor: child.props.htmlFor || inputId,
+            id: child.props.id || labelId,
+          }));
+        }
+
+        if (child.type === StatusMessage) {
+          return cloneElementWithRef(child, (child) => ({
+            ...child.props,
+            id: child.props.id || messageId,
+          }));
+        }
+
+        if (
+          isInput(child) ||
+          child.type === InputWithDecorations ||
+          child.type === InputWithIcon
+        ) {
+          return handleCloningInputs(child, {
+            labelId,
+            inputId,
+            messageId,
+          });
+        }
+
+        return child;
+      }),
+    [children, inputId, labelId, messageId],
+  );
+};
+
+//-------------------------------------------------------------------------------
+
+/**
+ * Setup/prerequisite for `useChildrenWithIds` to gather information from children.
+ *
+ * @returns the following ids (prefers id from props, otherwise generates one)
+ *  - `labelId`
+ *  - `inputId`
+ *  - `messageId`
+ */
+const useSetup = (children: React.ReactNode) => {
+  const idPrefix = useId();
+
+  let labelId: string | undefined;
+  let inputId: string | undefined;
+  let messageId: string | undefined;
+
+  let hasLabel = false;
+  let hasSelect = false;
+
+  const findInputId = (child: React.ReactNode) => {
+    if (!React.isValidElement(child)) {
+      return;
+    }
+    if (child.type === ComboBox) {
+      return child.props.inputProps?.id || `${idPrefix}--input`;
+    } else if (child.type !== Select) {
+      return child.props.id || `${idPrefix}--input`;
+    }
+  };
+
+  React.Children.forEach(children, (child) => {
+    if (!React.isValidElement(child)) {
+      return;
+    }
+
+    if (child.type === Label || child.type === 'label') {
+      hasLabel = true;
+      labelId ||= child.props.id || `${idPrefix}--label`;
+    }
+
```

Comment:
what does this have to do with Select?

Reply:
that's also unrelated, you might be confusing two different parts:

- L185: We only want `labelId` for `Select`, because it gets referenced in `aria-labelledby` (other inputs use `<Label htmlFor`>)
- L150: We exclude `Select` because its `id` gets set on the wrapping `<div>`. To get its real id, we would have to grab it from `triggerProps`, similar to how we do for ComboBox (L149), but it doesn't matter because we don't even return `inputId` for Select (L186).

Reply:
added some comments [`ab5c823`](https://github.com/iTwin/iTwinUI/pull/1777/commits/ab5c823f3e3e477b3eef6e0ad6618cf827c6f607)